src/main/scala/org/apache/spark/sql/TiStrategy.scala

```diff
@@ -141,68 +142,67 @@ class TiStrategy(context: SQLContext) extends Strategy with Logging {
+    import scala.collection.JavaConverters._
+
+    val tiFilters = filters.collect { case BasicExpression(expr) => expr }.asJava
     val scanBuilder: ScanBuilder = new ScanBuilder
     val scanPlan = if (allowIndexDoubleRead()) {
-      scanBuilder.buildScan(JavaConversions.seqAsJavaList(tiFilters), source.table)
+      scanBuilder.buildScan(tiFilters, source.table)
     } else {
-      scanBuilder.buildTableScan(JavaConversions.seqAsJavaList(tiFilters), source.table)
+      scanBuilder.buildTableScan(tiFilters, source.table)
     }
 
     dagRequest.addRanges(scanPlan.getKeyRanges)
-    scanPlan.getFilters.toList.map(dagRequest.addWhere)
+    scanPlan.getFilters.asScala.foreach(dagRequest.addWhere)
     if (scanPlan.isIndexScan) {
       dagRequest.setIndexInfo(scanPlan.getIndex)
     }
     dagRequest
   }
 
-  def addSortOrder(request: TiDAGRequest, sortOrder: Seq[SortOrder]): Unit =
-    if (sortOrder != null) {
-      sortOrder.foreach(
-        (order: SortOrder) =>
-          request.addOrderByItem(
-            TiByItem.create(
-              BasicExpression.convertToTiExpr(order.child).get,
-              order.direction.sql.equalsIgnoreCase("DESC")
-            )
+  private def addSortOrder(request: TiDAGRequest, sortOrder: Seq[SortOrder]): Unit =
+    sortOrder.foreach { order: SortOrder =>
+      request.addOrderByItem(
+        TiByItem.create(
+          BasicExpression.convertToTiExpr(order.child).get,
+          order.direction.sql.equalsIgnoreCase("DESC")
         )
       )
     }
 
-  def pruneTopNFilterProject(
+  private def pruneTopNFilterProject(
     limit: Int,
     projectList: Seq[NamedExpression],
     filterPredicates: Seq[Expression],
     source: TiDBRelation,
-    sortOrder: Seq[SortOrder] = null
+    sortOrder: Seq[SortOrder]
   ): SparkPlan = {
     val request = new TiDAGRequest(pushDownType(), timeZoneOffset())
     request.setLimit(limit)
     addSortOrder(request, sortOrder)
     pruneFilterProject(projectList, filterPredicates, source, request)
   }
 
-  def collectLimit(limit: Int, child: LogicalPlan): SparkPlan = child match {
+  private def collectLimit(limit: Int, child: LogicalPlan): SparkPlan = child match {
     case PhysicalOperation(projectList, filters, LogicalRelation(source: TiDBRelation, _, _))
         if filters.forall(TiUtils.isSupportedFilter(_, source, blacklist)) =>
-      pruneTopNFilterProject(limit, projectList, filters, source, null)
+      pruneTopNFilterProject(limit, projectList, filters, source, Nil)
     case _ => planLater(child)
   }
 
-  def takeOrderedAndProject(
+  private def takeOrderedAndProject(
     limit: Int,
     sortOrder: Seq[SortOrder],
     child: LogicalPlan,
     project: Seq[NamedExpression]
   ): SparkPlan = {
-    // If sortOrder is not null, limit must be greater than 0
-    if (limit < 0 || (sortOrder == null && limit == 0)) {
+    // If sortOrder is empty, limit must be greater than 0
+    if (limit < 0 || (sortOrder.isEmpty && limit == 0)) {
```

Comment:
Did I miss something here? I don't think `SortOrder` can be null in Spark.

Reply:
Sorry it's my fault, originally `takeOrderedAndProject` is called in `collectLimit`, and an explicit null sortOrder will be passed, I removed that call in `collectLimit` but forgot to remove these checks here, thanks for your dedicated review!